### PR TITLE
Non-blocking PUT in CHPL_COMM=ofi

### PR DESCRIPTION
**Pull request by jhh67**
This PR implements non-blocking PUTs in `CHPL_COMM=ofi` correctly. Prior to Chapel 2.2, non-blocking PUTs were implemented via blocking PUTs, and blocking PUTs would be typically be "injected". The OFI injection functionality allows the source buffer to be reused immediately, and suppresses the completion event that normally indicates that the operation is complete. As a result, both blocking and non-blocking PUTs would incorrectly be considered complete when the function returned. Because injection allows the source buffer to be reused immediately this would not cause a race on the source buffer, but it would allow PUTs to be reordered and could potentially cause hangs because the upper layers would not know that there were outstanding PUTs that might require progressing the communication endpoint.  Although it doesn't appear that we saw these problems in production use, the potential existed.

This PR flips the situation around so that a blocking PUT is implemented as a non-blocking PUT followed by waiting for the PUT to complete. The lower-level communication primitives are now all non-blocking. PUTs may still be injected, but I will likely remove that functionality in a subsequent PR because 1) it doesn't make sense to inject a blocking PUT because you are going to wait for it to complete anyway, and 2) the semantics of non-blocking PUTs in Chapel are that the caller may not reuse the source buffer until the PUT completes. As a result, neither benefit from injection.

As part of implementing this PR I changed the dangling PUT logic. A "dangling PUT" one that may not be visible, and will need to be forced into visibility when required by the Chapel memory consistency model. Previously, PUTs were not allowed to dangle if transmit contexts were unbound, which led to a lot of special-case code to deal with bound and unbound transmit contexts. Now all dangling PUTs are treated the same whether or not the transmit contexts are bound, and dangling PUTs are forced into visibility when an unbound transmit context is freed. They could possibly be deferred until even later, but we expect unbound transmit contexts to be the exception so it doesn't make sense to optimize them further. This change simplifies the code without degrading performance.

### CODE REVIEW

File: test/runtime/configMatters/comm/large-rma/EXECENV
Comment (jabraham17) [inline test/runtime/configMatters/comm/large-rma/EXECENV:1]: just a nit that this file is missing its newline

File: test/runtime/configMatters/comm/large-rma/README
Comment (jabraham17) [inline test/runtime/configMatters/comm/large-rma/README:1]: same nit, this file is missing its newline

File: test/runtime/configMatters/comm/unbound/EXECENV
Comment (jabraham17) [inline test/runtime/configMatters/comm/unbound/EXECENV:1]: same nit, missing newline

File: test/runtime/configMatters/comm/unbound/README
Comment (jabraham17) [inline test/runtime/configMatters/comm/unbound/README:1]: same nit, missing newline

File: test/runtime/configMatters/comm/unbound/SKIPIF
Comment (jabraham17) [inline test/runtime/configMatters/comm/unbound/SKIPIF:1]: same nit, missing newline

File: runtime/src/comm/ofi/comm-ofi.c [hunk 1]
```diff
@@ -5446,13 +5510,87 @@ void amCheckLiveness(void) {
...
+  CHK_TRUE(raddr != NULL);
+
+  if (size == 0) {
+    goto done;
+  }
+
+  if (node == chpl_nodeID) {
+    memmove(raddr, addr, size);
+    goto done;
+  }
```
Comment (jabraham17) [inline runtime/src/comm/ofi/comm-ofi.c:5552]: I know we already talked about this offline, but this kind of "goto done" pattern does not sit right with me. I understand the usefulness of having the cleanup code all done in one place that is jumped to, making it harder to forget to call the cleanup code (leading to leaks, etc). But in this case there is no cleanup, just "return proceed". 

Which to my mind suggests the following "simple fix"
1. remove the goto and inline "return proceed"
2. remove "proceed" altogether, and just make it "return false" and "return true
I think this change will improve readability.

However, this is arguably still not great and I know you expressed a strong preference for a single return point per function. So consider restructuring as
```
if size == 0
  do nothing
  proceed is false
else if node is current node
  do memove
  proceed is false
else 
  put_prologue_inner
  proceed is true
return proceed
```

Then all the logic for the prologue goes into the `_inner` method (or `_helper` or whatever descriptive name makes sense)

The same note for `check_complete`

Comment (jabraham17) [reply runtime/src/comm/ofi/comm-ofi.c:5552]: As a final note, if extra function overhead is a concern I would suggest just using `static inline`, or use `static ___always_inline` (defined in sys_basic.h) if you really want it inlined

**Review by jabraham17 — APPROVED** [review]

File: runtime/src/comm/ofi/comm-ofi.c (hunk 1 above)
Comment (jhh67) [reply runtime/src/comm/ofi/comm-ofi.c:5552]: Done.

**Comment by bradcray** [comment]
@jhh67 : In your table above, would Chapel 2.2 be expected to essentially match the `main` column?

**Comment by jhh67** [comment]
@bradcray: Yes.

**Pull request merged**